### ISSUE TITLE
How to make 2D arrowhead points point to endpoint instead of being centered at endpoint?

### ISSUE THREAD

**Issue by khiner** [MakieOrg/Makie.jl#2607]
I am having trouble accounting for the fact that Makie's `arrows!` function _centers_ arrowheads at the vector endpoints instead of _pointing_ to the endpoint.

Here is my code with default behavior:

```julia
arrowsize=40
# For each arrow...
arrows!([origin.x], [origin.y], [val.x], [val.y]; arrowsize=arrowsize, ...)
```
<img width="692" alt="Screen Shot 2023-01-17 at 11 01 07 AM" src="https://user-images.githubusercontent.com/685394/212950921-70707150-0fd4-4daf-9fbd-1b8aab4af7a4.png">

And here is my naive attempt at adjusting for arrowhead size:

```julia
arrowsize = 40 # In pixel space
pix_scale_2d = camera(Scene()).pixel_space[][1:2,1:2]
# For each arrow...
# Convert scene value to pix value.
val_pixels = (pix_scale_2d ^ -1) * val
# Subtract half the arrow in pixel space, in the direction of the value.
val_pixels -= normalize(val) * (arrowsize/2)
# Convert value back to pixel space.
val = pix_scale_2d * val_pixels
arrows!([origin.x], [origin.y], [val[1]], [val[2]]; arrowsize=arrowsize, ...)
```
<img width="667" alt="Screen Shot 2023-01-17 at 12 27 43 PM" src="https://user-images.githubusercontent.com/685394/212969779-0c2830b0-76ba-4119-9f51-cc4a5a0d0ce8.png">

If I subtract  `normalize(val) * (arrowsize*2)` instead of `normalize(val) * (arrowsize/2)` (which in my head is twice the arrow size instead of half, but is clearly not in actuality), I get:

<img width="674" alt="Screen Shot 2023-01-17 at 12 27 57 PM" src="https://user-images.githubusercontent.com/685394/212970096-71883728-5e69-4eb8-b835-f1773eeb8fc1.png">

It seems like my naive attempt is pretty close, but most likely for the wrong reasons. What am I doing wrong here?

Thanks for the help!

**Comment by khiner** [comment]
Also, I would very much appreciate an `arrow` option to make this arrow end at the vector point. Ideally, this would be the default, like it is in the Plots package. I would think most use cases would prefer arrows ending exactly where the vector ends rather than placing an arrow marker centered around the vector end point.

**Comment by jkrumbiegel** [comment]
Yeah I think the only reason is that it was simpler to implement at the time, and arrows was meant for showing directions, not for pointing at specific points. That's also why it takes a direction vector, not a target point, because that's common for specifying vector fields.

I was thinking about writing a more flexible recipe for annotation arrows, where you can also have curved paths and the arrow always points at the target (or slightly before it), etc.

**Comment by khiner** [comment]
Thank you for the quick response.

> You cannot solve the problem in data space because the arrow heads have a size independent of data scale. That's why I said the other implementation was easier.

Sorry - with "other implementation," are you referring to the potential more flexible recipe (to come down the line) or a potential workaround that I could use (in which case I'm not seeing in your response)?

**Comment by ffreyer** [comment]
You might have more luck with your approach if you define your own `arrowhead`. The default `:utriangle` does not fill a markersize x markersize square and neither does `BezierUTriangle`. That's probably throwing of the math. This would be space filling:

```julia
f, a, p = scatter(Point2f(0), markersize = 100, marker = Rect, color = :black)
tri = BezierPath([
    MoveTo(Point2f(-0.5, -0.5)), LineTo(0, 0.5), LineTo(0.5, -0.5), ClosePath()
])
scatter!(a, Point2f(0), markersize = 100, marker = tri, color = :red)
f
```
![Screenshot from 2023-01-17 18-58-47](https://user-images.githubusercontent.com/10947937/212975934-c3359038-7e4f-4abf-a384-cbd9aa61cc47.png)



Related: #2472

**Comment by jkrumbiegel** [comment]
I think for his purposes, he should use 
```julia
tri = BezierPath([
    MoveTo(Point2f(-0.5, -1)), LineTo(0, 0), LineTo(0.5, -1), ClosePath()
])
```
or so because then the arrow tip would be at the origin. However, this is not going to look perfect because the end of the line will "blunt" the arrow tip, showing through (I assume, didn't test).

By "other" implementation I meant the one we have, other relative to the one I'm planning.

**Comment by khiner** [comment]
Thank you for the help! @jkrumbiegel this `BezierPath` approach (as a marker) will only work with CairoMakie, right?

> By "other" implementation I meant the one we have, other relative to the one I'm planning.

👍 got it, thanks.

**Comment by khiner** [comment]
Next issue I ran into looks like this one (marker `rotations` not taking axis ratio into account): https://github.com/MakieOrg/Makie.jl/issues/860

```julia
scatter!(
  [o + v.value];
  marker=tri,
  rotations=[-pi/2 + atan(vec.y, vecx)],
)
```

If I manually scale my `vec` with `pix_scale_2d * vec`, the angle looks right.

**Comment by jkrumbiegel** [comment]
no BezierPath markers work in GLMakie as well, they are written to little textures first, like text glyphs. And yes, to correct rotation for axis scaling you need to take the projection into account, I think the `arrows` recipe does that and you could compare with it.
